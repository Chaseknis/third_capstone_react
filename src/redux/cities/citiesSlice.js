import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import getWeatherData from '../api/api';

export const fetchCities = createAsyncThunk(
  'cities/fetchCities',
  async (cityNames) => {
    const citiesData = await Promise.all(
      cityNames.map(async (cityName) => {
        const weatherData = await getWeatherData(cityName);
        return { cityName, ...weatherData };
      }),
    );
    return citiesData;
  },
);

export const citiesSlice = createSlice({
  name: 'cities',
  initialState: {
    citiesData: [],
    status: 'idle',
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchCities.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(fetchCities.fulfilled, (state, action) => ({
        ...state,
        status: 'succeeded',
        citiesData: action.payload,
      }))
      .addCase(fetchCities.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        error: action.error.message,
      }));
  },
});

export const selectCitiesData = (state) => state.cities.citiesData;
export const selectCitiesStatus = (state) => state.cities.status;
export const selectCitiesError = (state) => state.cities.error;

export default citiesSlice.reducer;
