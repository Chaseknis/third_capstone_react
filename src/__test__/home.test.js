import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchCities } from '../redux/cities/citiesSlice';
import Home from '../routers/home';

jest.mock('react-redux');

describe('Home component', () => {
  const cities = [
    'paris',
    'london',
    'new york',
    'tokyo',
    'new jersey',
    'birmingham',
    'los angeles',
    'sydney',
    'toronto',
    'manchester',
  ];

  const weatherData = [
    { name: 'Paris', temp: 15 },
    { name: 'London', temp: 12 },
    { name: 'New York', temp: 10 },
    { name: 'Tokyo', temp: 18 },
    { name: 'New Jersey', temp: 8 },
    { name: 'Birmingham', temp: 11 },
    { name: 'Los Angeles', temp: 22 },
    { name: 'Sydney', temp: 25 },
    { name: 'Toronto', temp: 5 },
    { name: 'Manchester', temp: 9 },
  ];

  beforeEach(() => {
    useSelector.mockImplementation((selector) => selector({
      citiesData: weatherData,
    }));
  });

  afterEach(() => {
    useSelector.mockClear();
  });

  test('renders the header and small title', () => {
    render(<Home />);

    expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent('Weather App');
    expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent('Forecast by City');
  });

  test('renders a card for each city', () => {
    render(<Home />);

    const cards = screen.getAllByRole('button');

    expect(cards).toHaveLength(cities.length);

    cards.forEach((card, index) => {
      expect(card).toHaveAttribute('href', '/details');
      expect(card).toHaveTextContent(cities[index]);
    });
  });

  test('sets the cityname in localStorage on card click', () => {
    render(<Home />);

    const cards = screen.getAllByRole('button');

    cards.forEach((card, index) => {
      fireEvent.click(card);
      expect(localStorage.setItem).toHaveBeenLastCalledWith('cityname', JSON.stringify(weatherData[index].name));
    });
  });

  test('dispatches the fetchCities action on mount', () => {
    const dispatch = jest.fn();
    useDispatch.mockReturnValue(dispatch);

    render(<Home />);

    expect(dispatch).toHaveBeenCalledWith(fetchCities(cities));
  });

  test('matches snapshot', () => {
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });
});
