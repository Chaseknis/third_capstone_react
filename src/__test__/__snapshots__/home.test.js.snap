// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details component Details component 1`] = `
<main>
  <div
    className="home_wrapper"
  >
    <div
      className="header_wrapper p-r-10"
    >
      <h1>
        Weather App
      </h1>
    </div>
    <h2
      className="small_title p-r-10"
    >
      Forecast by City
    </h2>
    <div
      className="filter_wrapper"
    >
      <input
        onChange={[Function]}
        placeholder="Enter City Name"
        type="text"
        value=""
      />
    </div>
    <div
      className="cards_wrapper"
    />
  </div>
</main>
`;
